File: app/src/main/java/com/afirez/knight/aop/AppAspect.java
package com.afirez.knight.aop;

import android.util.Log;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;

@Aspect
public class AppAspect {

    @Around("execution( * android.app.Application.onCreate(..))")
    public void aopAppOnCreateAdvice(ProceedingJoinPoint joinPoint) throws Throwable {
        Log.i("aspectj", String.format("------------>> Application.onCreate: \n%s \n%s",
                joinPoint.getSignature(), joinPoint.getThis()));
        joinPoint.proceed();
        Log.i("aspectj", "<<------------ Application.onCreate");
    }

    @Around("execution( * android.support.v4.app.FragmentActivity.onCreate(..))")
    public void aopActivityOnCreateAdvice(ProceedingJoinPoint joinPoint) throws Throwable {
        Log.i("aspectj", String.format("------------>> Activity.onCreate: \n%s \n%s \n%s",
                joinPoint.getSignature(), joinPoint.getThis(), joinPoint.getArgs()[0]));
        joinPoint.proceed();
        Log.i("aspectj", "<<------------ Activity.onCreate");
    }

    @Around("execution( * android.support.v4.app.Fragment.onCreate(..))")
    public void aopFragmentOnCreateAdvice(ProceedingJoinPoint joinPoint) throws Throwable {
        Log.i("aspectj", String.format("------------>> Fragment.onCreate: \n%s \n%s \n%s",
                joinPoint.getSignature(), joinPoint.getThis(), joinPoint.getArgs()[0]));
        joinPoint.proceed();
        Log.i("aspectj", "<<------------ Fragment.onCreate");
    }

}

File: lib/src/main/java/com/afirez/lib/aop/AppAspect.kt
package com.afirez.lib.aop

import android.util.Log

import com.afirez.lib.User
import com.afirez.lib.lib
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.kodein.di.Kodein
import org.kodein.di.KodeinAware
import org.kodein.di.generic.instance

@Aspect
class AppAspect : KodeinAware {
    override val kodein: Kodein = Kodein.lazy {
        import(lib)
//        bind<User>(tag = "user") with singleton { User() }
    }

    val user: User by instance(tag = "user")

    @Around("execution( * android.app.Application.onCreate(..))")
    @Throws(Throwable::class)
    fun aopAppOnCreateAdvice(joinPoint: ProceedingJoinPoint) {
        Log.i("aspectk", "------------>> Application.onCreate: \n${joinPoint.signature} \n${joinPoint.`this`}")
        joinPoint.proceed()
        Log.i("aspectk", "<<------------ Application.onCreate ${user.name}")
    }

    @Around("execution( * android.support.v4.app.FragmentActivity.onCreate(..))")
    @Throws(Throwable::class)
    fun aopActivityOnCreateAdvice(joinPoint: ProceedingJoinPoint) {
        Log.i("aspectk", "------------>> Activity.onCreate: \n${joinPoint.signature} \n${joinPoint.`this`} \n${joinPoint.args[0]}")
        joinPoint.proceed()
        Log.i("aspectk", "<<------------ Activity.onCreate")
    }

    @Around("execution( * android.support.v4.app.Fragment.onCreate(..))")
    @Throws(Throwable::class)
    fun aopFragmentOnCreateAdvice(joinPoint: ProceedingJoinPoint) {
        Log.i("aspectk", "------------>> Fragment.onCreate: \n${joinPoint.signature} \n${joinPoint.`this`} \n${joinPoint.args[0]}")
        joinPoint.proceed()
        Log.i("aspectk", "<<------------ Fragment.onCreate")
    }

}